README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

:four_leaf_clover: Package help show popup reaction like Facebook and wrapper for widget (Show anyone who reaction this message).

## :four_leaf_clover::four_leaf_clover: Preview
<p>
<img src="https://github.com/Askany-NDN/reaction_button_askany/blob/main/screenshots/IMG_0470.png?raw=true" width=180/>
<img src="https://github.com/Askany-NDN/reaction_button_askany/blob/main/screenshots/IMG_0471.png?raw=true" width=180/>
<img src="https://github.com/Askany-NDN/reaction_button_askany/blob/main/screenshots/IMG_0465.png?raw=true" width=180/>
<img src="https://github.com/Askany-NDN/reaction_button_askany/blob/main/screenshots/IMG_0467.png?raw=true" width=180/>
</p>

## Features

* Support show popup emotions picker
* Support reaction wrapper for message card use case

## Usage
- Show reaction popup

```dart
        GestureDetector(
          onTapDown: (details) {
            ReactionAskany.showReactionBox(
              context,
              offset: details.globalPosition,
              boxParamenters: boxParamenters,
              emotionPicked: _emotion,
              handlePressed: (Emotions emotion) {
                setState(() {
                  _emotion = emotion;
                });
              },
            );
          },
          child: widget.buttonReaction,
        ),
```
    
- Use reaction wrapper - useful for reaction message
    
```dart
            ReactionWrapper(
              boxParamenters: ReactionBoxParamenters(
                brightness: Brightness.light,
                iconSize: 26,
                iconSpacing: 10,
                paddingHorizontal: 30,
                radiusBox: 40,
                quantityPerPage: 6,
              ),
              buttonReaction: const Padding(
                padding: EdgeInsets.only(top: 2.0),
                child: Icon(
                  Icons.face_outlined,
                  size: 20.0,
                  color: Colors.grey,
                ),
              ),
              child: Container(
                padding: const EdgeInsets.symmetric(
                  vertical: 12.0,
                  horizontal: 20.0,
                ),
                decoration: BoxDecoration(
                  color: Colors.greenAccent.shade100,
                  borderRadius: BorderRadius.circular(30),
                ),
                child: const Text(
                  'Message from lambiengcode <3',
                ),
              ),
            ),
```

### ReactionBoxParamenters
| parameter                  | description                                                                           | default                                                                                                                                                                               |
| -------------------------- | ------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| iconSize          | Size of emotion in reaction box                                           |20|
| iconSpacing                  | Padding horizontal value for each emotion |8|
| paddingHorizontal            | Padding horizontal value for reaction box                                                        | 16 |
| radiusBox               | Radius circular of reaction box                             | 10 |
| quantityPerPage      | Number of emotions per page view                           | 5 |
| brightness                  | Set brightness for show background color compatitive                                  | Brightness.light
## Download Askany

<p>
<a href="https://apps.apple.com/vn/app/askany/id1589217505"><img src="https://askany.com/images/app-store.png" height="60px" width="180px"/></a>
<a href="https://play.google.com/store/apps/details?id=com.askany"><img src="https://askany.com/images/ch-play.png" height="60px" width="180px"/></a>
</p>

## License - lambiengcode

```terminal
MIT License

Copyright (c) 2022 Askany

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

```


CHANGELOG.md
## 0.1.2

* [Fix] position showMenu reaction
* Add double tap message for open reaction menu

## 0.1.0

* Update homepage and repository

## 0.0.3

* Fix not excute handlePressed in ReactionWrapper
* Add animation when choose emotion
* Choose current emotion will remove reaction

## 0.0.2

* Allow pressed emotion for view who react this message
* Padding child in ReactionWrapper when has emotion
## 0.0.1

* Support show popup emotions picker
* Support reaction wrapper for message card use case


lib/reaction_askany.dart
library reaction_askany;

export 'widgets/reaction_wrapper.dart';

import 'package:flutter/material.dart';
import 'package:reaction_askany/models/emotions.dart';
import 'package:reaction_askany/models/reaction_box_paramenters.dart';
import 'package:reaction_askany/widgets/reaction_box.dart';

class ReactionAskany {
  static void showReactionBox(
    BuildContext context, {
    required Offset offset,
    Function(Emotions)? handlePressed,
    Emotions? emotionPicked,
    ReactionBoxParamenters? boxParamenters,
    String? doubleTapLabel,
  }) async {
    final double left = offset.dx;
    final double top = offset.dy;

    final ReactionBoxParamenters paramenters =
        boxParamenters ?? ReactionBoxParamenters();

    await showMenu(
      context: context,
      color: Colors.transparent,
      position: RelativeRect.fromLTRB(
        left,
        top - paramenters.reactionBoxHeight - paramenters.iconSpacing * 2.5,
        left,
        top - paramenters.iconSpacing * 3,
      ),
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(
          paramenters.radiusBox,
        ),
      ),
      constraints: BoxConstraints(
        maxWidth: paramenters.reactionBoxWidth,
      ),
      items: [
        PopupMenuItem(
          padding: EdgeInsets.zero,
          enabled: false,
          onTap: () {},
          child: Container(
            decoration: BoxDecoration(
              border: Border.all(
                color: Brightness.light == paramenters.brightness
                    ? Colors.grey.shade300
                    : Colors.grey.shade800,
                width: .2,
              ),
              color: Brightness.light == paramenters.brightness
                  ? Colors.white
                  : const Color(0xFF22222A),
              borderRadius: BorderRadius.circular(
                paramenters.radiusBox,
              ),
            ),
            height: paramenters.reactionBoxHeight + 10.0,
            width: paramenters.reactionBoxWidth,
            padding: const EdgeInsets.symmetric(
              vertical: 4.0,
            ),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Visibility(
                  visible: doubleTapLabel != null,
                  child: Padding(
                    padding: const EdgeInsets.symmetric(vertical: 3.0),
                    child: Text(
                      doubleTapLabel!,
                      textAlign: TextAlign.center,
                      style: const TextStyle(
                        fontSize: 9.5,
                      ),
                    ),
                  ),
                ),
                ReactionBox(
                  emotions: Emotions.values,
                  handlePressed: handlePressed ?? (Emotions emo) {},
                  boxParamenters: paramenters,
                  emotionPicked: emotionPicked,
                ),
              ],
            ),
          ),
        ),
      ],
      elevation: 0,
    );
  }
}


lib/widgets/reaction_box.dart
import 'package:flutter/material.dart';
import 'package:reaction_askany/models/emotions.dart';
import 'package:reaction_askany/models/reaction_box_paramenters.dart';
import 'package:reaction_askany/widgets/reaction_page.dart';

class ReactionBox extends StatefulWidget {
  final List<Emotions> emotions;
  final Function(Emotions) handlePressed;
  final ReactionBoxParamenters boxParamenters;
  final Emotions? emotionPicked;

  const ReactionBox({
    super.key,
    required this.emotions,
    required this.handlePressed,
    required this.boxParamenters,
    required this.emotionPicked,
  });

  @override
  State<StatefulWidget> createState() => _ReactionBoxState();
}

class _ReactionBoxState extends State<ReactionBox> {
  late PageController _pageController;
  final List<List<Emotions>> _emotions = [];
  int _currentIndex = 0;

  @override
  void initState() {
    super.initState();
    _pageController = PageController(
      initialPage: _currentIndex,
    );

    List<Emotions> temp = [];
    for (Emotions emotion in widget.emotions) {
      temp.add(emotion);
      if (temp.length == widget.boxParamenters.quantityPerPage) {
        _emotions.add(temp);
        temp = [];
      }
    }

    if (temp.isNotEmpty) {
      _emotions.add(temp);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.center,
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Container(
          height: widget.boxParamenters.iconSize +
              widget.boxParamenters.iconSpacing * 2,
          width: double.infinity,
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(
              widget.boxParamenters.radiusBox,
            ),
            color: Colors.transparent,
          ),
          alignment: Alignment.center,
          child: PageView(
            onPageChanged: (int index) {
              setState(() {
                _currentIndex = index;
              });
            },
            physics: const ClampingScrollPhysics(),
            controller: _pageController,
            children: _emotions
                .map<Widget>(
                  (page) => ReactionPage(
                    emotions: page,
                    handlePressed: widget.handlePressed,
                    boxParamenters: widget.boxParamenters,
                    emotionPicked: widget.emotionPicked,
                  ),
                )
                .toList(),
          ),
        ),
        Visibility(
          visible: _emotions.length > 1,
          child: SizedBox(
            height: 10.0,
            child: ListView.builder(
              padding: EdgeInsets.zero,
              shrinkWrap: true,
              scrollDirection: Axis.horizontal,
              physics: const NeverScrollableScrollPhysics(),
              itemCount: _emotions.length,
              itemBuilder: ((context, index) {
                final bool isCurrent = index == _currentIndex;
                return Container(
                  margin: const EdgeInsets.only(right: 4),
                  height: isCurrent ? 6 : 4,
                  width: isCurrent ? 6 : 4,
                  decoration: BoxDecoration(
                    color: isCurrent ? Colors.grey : Colors.grey.shade400,
                    shape: BoxShape.circle,
                  ),
                );
              }),
            ),
          ),
        ),
      ],
    );
  }
}


lib/widgets/reaction_page.dart
import 'package:flutter/material.dart';
import 'package:reaction_askany/models/emotions.dart';
import 'package:reaction_askany/models/reaction_box_paramenters.dart';
import 'package:reaction_askany/widgets/emotion_widget.dart';

class ReactionPage extends StatelessWidget {
  final List<Emotions> emotions;
  final Function(Emotions) handlePressed;
  final ReactionBoxParamenters boxParamenters;
  final Emotions? emotionPicked;

  const ReactionPage({
    super.key,
    required this.emotions,
    required this.handlePressed,
    required this.boxParamenters,
    required this.emotionPicked,
  });

  @override
  Widget build(BuildContext context) {
    return ListView.builder(
      padding: EdgeInsets.symmetric(
        horizontal: boxParamenters.paddingHorizontal,
      ),
      scrollDirection: Axis.horizontal,
      shrinkWrap: true,
      physics: const NeverScrollableScrollPhysics(),
      itemCount: emotions.length,
      itemBuilder: (context, index) {
        return EmotionWidget(
          emotion: emotions[index],
          handlePressed: handlePressed,
          boxParamenters: boxParamenters,
          isSelected: emotions[index] == emotionPicked,
        );
      },
    );
  }
}


lib/widgets/emotion_widget.dart
import 'package:flutter/material.dart';
import 'package:reaction_askany/models/emotions.dart';
import 'package:reaction_askany/models/reaction_box_paramenters.dart';

class EmotionWidget extends StatefulWidget {
  final Emotions emotion;
  final Function(Emotions) handlePressed;
  final ReactionBoxParamenters boxParamenters;
  final bool isSelected;

  const EmotionWidget({
    super.key,
    required this.emotion,
    required this.handlePressed,
    required this.boxParamenters,
    this.isSelected = false,
  });

  @override
  State<EmotionWidget> createState() => _EmotionWidgetState();
}

class _EmotionWidgetState extends State<EmotionWidget>
    with TickerProviderStateMixin {
  late AnimationController _controller;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      duration: const Duration(milliseconds: 1500),
      vsync: this,
    );
    if (widget.isSelected) {
      _controller.repeat();
    }
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        widget.handlePressed(widget.emotion);
        Navigator.pop(context);
      },
      child: widget.isSelected
          ? Container(
              decoration: BoxDecoration(
                color: Colors.grey.shade800,
                borderRadius: BorderRadius.circular(8.0),
              ),
              child: ScaleTransition(
                scale: Tween(begin: 1.0, end: 1.1).animate(
                  CurvedAnimation(
                    curve: Curves.easeInOut,
                    parent: _controller,
                  ),
                ),
                child: _buildBody(),
              ),
            )
          : _buildBody(),
    );
  }

  Widget _buildBody() {
    return Container(
      color: Colors.transparent,
      padding: EdgeInsets.symmetric(
        horizontal: widget.boxParamenters.iconSpacing,
      ),
      alignment: Alignment.center,
      child: Image.asset(
        widget.emotion.assetImage,
        height: widget.boxParamenters.iconSize,
        width: widget.boxParamenters.iconSize,
      ),
    );
  }
}


lib/widgets/reaction_wrapper.dart
import 'package:flutter/material.dart';

import '../models/emotions.dart';
import '../models/reaction_box_paramenters.dart';
import '../reaction_askany.dart';

class ReactionWrapper extends StatefulWidget {
  final Widget child;
  final Widget buttonReaction;
  final Function(Emotions?)? handlePressed;
  final Function()? handlePressedReactions;
  final ReactionBoxParamenters? boxParamenters;
  final Emotions? initialEmotion;
  final String? doubleTapLabel;

  const ReactionWrapper({
    super.key,
    required this.child,
    required this.buttonReaction,
    this.boxParamenters,
    this.handlePressed,
    this.handlePressedReactions,
    this.initialEmotion,
    this.doubleTapLabel,
  });

  @override
  State<StatefulWidget> createState() => _ReactionWrapperState();
}

class _ReactionWrapperState extends State<ReactionWrapper>
    with TickerProviderStateMixin {
  final GlobalKey _reactionButtonKey = GlobalKey();
  final GlobalKey _messageWidgetKey = GlobalKey();
  late AnimationController _controller;
  Offset? _beginOffset;
  late final ReactionBoxParamenters boxParamenters =
      widget.boxParamenters ?? ReactionBoxParamenters();
  Emotions? _emotion;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      duration: const Duration(seconds: 1),
      vsync: this,
    );
    _emotion = widget.initialEmotion;
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.only(bottom: _emotion != null ? 12.0 : 0),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisAlignment: MainAxisAlignment.start,
        mainAxisSize: MainAxisSize.min,
        children: [
          Stack(
            clipBehavior: Clip.none,
            children: [
              GestureDetector(
                key: _messageWidgetKey,
                onDoubleTap: () async {
                  await _openReactionMenu(_messageWidgetKey);
                },
                child: widget.child,
              ),
              Visibility(
                visible: _emotion != null,
                child: Positioned(
                  bottom: -11,
                  child: GestureDetector(
                    onTap: widget.handlePressedReactions,
                    child: SlideTransition(
                      position: Tween<Offset>(
                        begin: _beginOffset ?? Offset.zero,
                        end: Offset.zero,
                      ).animate(CurvedAnimation(
                        parent: _controller,
                        curve: Curves.bounceOut,
                      )),
                      child: Container(
                        padding: const EdgeInsets.all(3.0),
                        margin: const EdgeInsets.only(
                          left: 20.0,
                        ),
                        decoration: BoxDecoration(
                          color: Brightness.light == boxParamenters.brightness
                              ? Colors.grey.shade100
                              : Colors.grey.shade800,
                          borderRadius: BorderRadius.circular(30.0),
                        ),
                        alignment: Alignment.center,
                        child: Image.asset(
                          _emotion?.assetImage ?? '',
                          height: 16,
                          width: 16,
                        ),
                      ),
                    ),
                  ),
                ),
              )
            ],
          ),
          const SizedBox(width: 4.0),
          Visibility(
            visible: _emotion == null,
            child: GestureDetector(
              key: _reactionButtonKey,
              onTapDown: (details) async {
                await _openReactionMenu(_reactionButtonKey);
              },
              child: widget.buttonReaction,
            ),
          ),
        ],
      ),
    );
  }

  Future<void> _openReactionMenu(GlobalKey widgetKey) async {
    setState(() {
      _beginOffset = Offset.zero;
    });

    if (FocusScope.of(context).hasFocus) {
      FocusScope.of(context).unfocus();
      await Future.delayed(const Duration(milliseconds: 300));
    }

    RenderBox? renderBox =
        widgetKey.currentContext?.findRenderObject() as RenderBox?;

    if (renderBox == null) return;

    Offset globalOffset = renderBox.localToGlobal(Offset.zero);

    // ignore: use_build_context_synchronously
    ReactionAskany.showReactionBox(
      context,
      offset: globalOffset,
      boxParamenters: boxParamenters,
      emotionPicked: _emotion,
      doubleTapLabel: widget.doubleTapLabel,
      handlePressed: (Emotions emotion) {
        _controller.reset();

        if (emotion == _emotion) {
          setState(() {
            _emotion = null;
          });

          if (widget.handlePressed != null) {
            widget.handlePressed!(_emotion);
          }

          return;
        }

        setState(() {
          _emotion = emotion;
          _beginOffset = const Offset(.25, -1);
        });

        _controller.forward();

        if (widget.handlePressed != null) {
          widget.handlePressed!(_emotion);
        }
      },
    );
  }
}


lib/models/reaction_box_paramenters.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'dart:convert';

import 'package:flutter/material.dart';

class ReactionBoxParamenters {
  final double iconSize;
  final double iconSpacing;
  final double paddingHorizontal;
  final double radiusBox;
  final int quantityPerPage;
  final Brightness brightness;
  ReactionBoxParamenters({
    this.iconSize = 20,
    this.iconSpacing = 8,
    this.paddingHorizontal = 16,
    this.radiusBox = 10,
    this.quantityPerPage = 5,
    this.brightness = Brightness.light,
  });

  ReactionBoxParamenters copyWith({
    double? iconSize,
    double? iconSpacing,
    double? paddingHorizontal,
  }) {
    return ReactionBoxParamenters(
      iconSize: iconSize ?? this.iconSize,
      iconSpacing: iconSpacing ?? this.iconSpacing,
      paddingHorizontal: paddingHorizontal ?? this.paddingHorizontal,
    );
  }

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'iconSize': iconSize,
      'iconSpacing': iconSpacing,
      'paddingHorizontal': paddingHorizontal,
    };
  }

  factory ReactionBoxParamenters.fromMap(Map<String, dynamic> map) {
    return ReactionBoxParamenters(
      iconSize: map['iconSize'] as double,
      iconSpacing: map['iconSpacing'] as double,
      paddingHorizontal: map['paddingHorizontal'] as double,
    );
  }

  String toJson() => json.encode(toMap());

  factory ReactionBoxParamenters.fromJson(String source) =>
      ReactionBoxParamenters.fromMap(
          json.decode(source) as Map<String, dynamic>);

  @override
  String toString() =>
      'ReactionBoxParamenters(iconSize: $iconSize, iconSpacing: $iconSpacing, paddingHorizontal: $paddingHorizontal)';

  @override
  bool operator ==(covariant ReactionBoxParamenters other) {
    if (identical(this, other)) return true;

    return other.iconSize == iconSize &&
        other.iconSpacing == iconSpacing &&
        other.paddingHorizontal == paddingHorizontal;
  }

  @override
  int get hashCode =>
      iconSize.hashCode ^ iconSpacing.hashCode ^ paddingHorizontal.hashCode;

  // Calculator
  double get reactionBoxHeight => iconSize + iconSpacing * 2 + 20.0;

  double get reactionBoxWidth =>
      (iconSize + iconSpacing * 2) * quantityPerPage + paddingHorizontal * 2;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:reaction_askany/models/reaction_box_paramenters.dart';
import 'package:reaction_askany/reaction_askany.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        color: Colors.white,
        width: double.infinity,
        padding: const EdgeInsets.symmetric(
          horizontal: 16.0,
        ),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            ReactionWrapper(
              boxParamenters: ReactionBoxParamenters(
                brightness: Brightness.light,
                iconSize: 26,
                iconSpacing: 10,
                paddingHorizontal: 30,
                radiusBox: 40,
                quantityPerPage: 6,
              ),
              buttonReaction: const Padding(
                padding: EdgeInsets.only(top: 2.0),
                child: Icon(
                  Icons.face_outlined,
                  size: 20.0,
                  color: Colors.grey,
                ),
              ),
              child: Container(
                padding: const EdgeInsets.symmetric(
                  vertical: 12.0,
                  horizontal: 20.0,
                ),
                decoration: BoxDecoration(
                  color: Colors.greenAccent.shade100,
                  borderRadius: BorderRadius.circular(30),
                ),
                child: const Text(
                  'Message from lambiengcode <3',
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
